idna/core.py

```diff
@@ -338,8 +338,11 @@ def uts46_remap(domain: str, std3_rules: bool = True, transitional: bool = False
 
 
 def encode(s: Union[str, bytes, bytearray], strict: bool = False, uts46: bool = False, std3_rules: bool = False, transitional: bool = False) -> bytes:
-    if isinstance(s, (bytes, bytearray)):
-        s = s.decode('ascii')
+    try:
+        if isinstance(s, (bytes, bytearray)):
+            s = s.decode('ascii')
+    except UnicodeDecodeError:
+        raise IDNAError('Invalid ASCII in A-label')
```

Comment:
Not a comment about this change, but I'm confused by this call to `.decode('ascii')` here. Isn't the point of `idna.encode()` here to transform non-ASCII characters into IDNA-encoded ASCII? If so this code path almost seems like we're maintaining it for backwards compatibility reasons instead of "functionality". Should our `IDNAError` here instead hint users to pass the value as a decoded string to force users to specify the encoding?

Reply:
I agree the error message would be more helpful if it simply said that you should be passing a unicode string to the function rather than a byte string.

Reply:
Also I would imagine that `if ... try ...` rather than `try ... if ...` would be more efficient for the vastly more common case that a unicode string is being passed.

Reply:
Agreed and modified as suggested